### ISSUE TITLE
gradient_accumulation_steps integration

### ISSUE THREAD

**Issue by stas00** [microsoft/DeepSpeed#758]
1. `gradient_accumulation_steps` configuration is not documented at all - it's only mentioned in the context of pipeline
2. there are no instructions on how to integrate it with the existing trainer - so if our trainer already handles the accumulation most likely we should reset `gradient_accumulation_steps` to 1 in deepspeed configuration, correct? I am asking since I don't know if perhaps something else is going on and DeepSpeed needs to know about it too.

To exemplify:
```
               if (step + 1) % self.args.gradient_accumulation_steps == 0 ...:
                    # Optimizer step
                    if self.deepspeed:
                        self.deepspeed.step()
              [...]
```

Thank you.

**Comment by stas00** [comment]
OK had a chance to read your code and investigate and found 4 invalid things in my initial integration https://github.com/microsoft/DeepSpeed/issues/776 Now need to think how to make the life of future integrator's easier.

The key was writing a test that compared weights and loss with 2 train runs with bs=8/grad_accum=1 and bs=4/grad_accum=2 and get the results to be identical.

**Comment by stas00** [comment]
So for future readers, the problem is that deepspeed does all the sorting out internally, so you basically don't need to implement any logic for gradient accumulation, unless like us you have to support various engines, so the correct implementation on our side was:
```
                # Optimizer step for deepspeed must be called on every step regardless of the value of gradient_accumulation_steps
                if self.deepspeed:
                    self.deepspeed.step()
                else:
                    if (step + 1) % self.args.gradient_accumulation_steps == 0 ...:
```

no gradient clipping either if you had one in that else branch. the correct solution was to remove deepspeed from most situations where you have some special logic - not even lr_scheduler!
